modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php

```diff
@@ -0,0 +1,236 @@
+<?php
+/**
+ * Background process runner class.
+ *
+ * @since n.e.x.t
+ * @package performance-lab
+ */
+
+/**
+ * Class Perflab_Background_Process.
+ *
+ * Runs the heavy lifting tasks in background in separate process.
+ */
+class Perflab_Background_Process {
+	/**
+	 * Name of the action which will trigger background
+	 * process to run the job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @var string Background process action name.
+	 */
+	const BG_PROCESS_ACTION = 'background_process_handle_request';
+
+	/**
+	 * Job instance.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @var Perflab_Background_Job Background job instance.
+	 */
+	private $job;
+
+	/**
+	 * Perflab_Background_Process constructor.
+	 *
+	 * Adds the necessary hooks to trigger process handling.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return void
+	 */
+	public function __construct() {
+		add_action( 'wp_ajax_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
+		add_action( 'wp_ajax_nopriv_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
+	}
+
+	/**
+	 * Handle incoming request to run the batch of job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @throws Exception Invalid request for background process.
+	 *
+	 * @return void
+	 */
+	public function handle_request() {
+		try {
```

Comment:
@ankitrox I feel like this method and try block does too much? I think we can split some of the logic between the `handle_request()` and `run()` methods. I think this because we have exceptions thrown that don't get caught as the job is not available.

The way I see it, `handle_request()` should verify and determine the request is valid. If not, the method should return early, or maybe even return a [`wp_send_json_error`](https://developer.wordpress.org/reference/functions/wp_send_json_error/) error message? If the request is valid it then calls the `run()` method,

Then the run method is responsible for running the job and I think this could contain everything from the `$this->job->lock()` line down.